pkg/reconciler/knativeserving/knativeserving.go

```diff
@@ -166,20 +182,67 @@ func (r *Reconciler) ensureFinalizerRemoval(_ context.Context, _ *mf.Manifest, i
 	return nil
 }
 
-// Delete obsolete resources from previous versions
-func (r *Reconciler) deleteObsoleteResources(ctx context.Context, manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing) error {
-	resources := []*unstructured.Unstructured{
-		// istio-system resources from 0.3.
-		common.NamespacedResource("v1", "Service", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("apps/v1", "Deployment", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("autoscaling/v1", "HorizontalPodAutoscaler", "istio-system", "knative-ingressgateway"),
-		// config-controller from 0.5
-		common.NamespacedResource("v1", "ConfigMap", instance.GetNamespace(), "config-controller"),
-	}
-	for _, r := range resources {
-		if err := manifest.Client.Delete(r); err != nil {
-			return err
-		}
+func (r *Reconciler) retrieveManifest(ctx context.Context, version string, instance *servingv1alpha1.KnativeServing) (mf.Manifest, error) {
+	if val, found := r.manifests[version]; found {
+		return val, nil
 	}
-	return nil
+
+	logger := logging.FromContext(ctx)
+	koDataDir := os.Getenv("KO_DATA_PATH")
+	manifesrDir := fmt.Sprintf("knative-serving/%s", version)
+	manifest, err := mf.NewManifest(filepath.Join(koDataDir, manifesrDir),
+		mf.UseClient(r.mfClient),
+		mf.UseLogger(zapr.NewLogger(logger.Desugar()).WithName("manifestival")))
+
+	if err != nil {
+		return manifest, err
+	}
+
+	if len(manifest.Resources()) == 0 {
+		return manifest, fmt.Errorf("unable to find the manifest for the Knative Serving version %s", version)
+	}
+
+	// Transform the manifest
+	transformers, err := r.transform(ctx, instance)
+	if err != nil {
+		return mf.Manifest{}, err
+	}
+
+	manifestTransformed, err := manifest.Transform(transformers...)
+	if err != nil {
+		return mf.Manifest{}, err
+	}
+
+	// Save the transformed manifest in the map
+	r.manifests[version] = manifestTransformed
+
+	return manifest, nil
 }
+
+func (r *Reconciler) getTargetManifest(ctx context.Context, instance *servingv1alpha1.KnativeServing) (mf.Manifest, error) {
+	if instance.Spec.GetVersion() != "" {
+		// If the version is set in spec of the CR, pick the version from the spec of the CR
+		return r.retrieveManifest(ctx, instance.Spec.GetVersion(), instance)
+	}
+
+	return r.getLatestManifest(ctx, instance)
```

Comment:
I believe this should return `r.getCurrentManifest(ctx, instance)` instead. Returning the latest violates the principle of least surprise, imo. We should have consensus on this and document it fiercely if this is the way we go. /cc @markusthoemmes @evankanderson

Reply:
@jcrossley3 @markusthoemmes @evankanderson 
I have some thoughts on how the auto-update could be implemented. Will open another PR.

The scenario is operator has installed Knative serving 0.15.0. There is a latest version 0.15.1 online available. Operator applies serving CR again, without setting the spec.version. Operator will TRY to find the `latest PATCH release` 0.15.1, and install it online. If it fails, it falls back to `r.getCurrentManifest(ctx, instance)`.

Reply:
I updated this section.
https://docs.google.com/document/d/1ehd5Yx4aSottuv7FYW4JeplNULJCs7nnDfy29y3NH8k/edit#heading=h.5502wzf6mqfe